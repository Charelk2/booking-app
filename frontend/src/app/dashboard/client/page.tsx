"use client";

import React, { useEffect, useState, useMemo } from "react";
import { useRouter, usePathname, useSearchParams } from "next/navigation";
import MainLayout from "@/components/layout/MainLayout";
import { useAuth } from "@/contexts/AuthContext";
import { Booking, BookingRequest } from "@/types";
import { getMyClientBookings, getMyBookingRequests } from "@/lib/api";
import SectionList from "@/components/dashboard/SectionList";
import BookingRequestCard from "@/components/dashboard/BookingRequestCard";
import { Spinner } from "@/components/ui";
import DashboardTabs from "@/components/dashboard/DashboardTabs";
import { format } from "date-fns";
import { formatCurrency, formatStatus } from "@/lib/utils";
import Link from "next/link";

export default function ClientDashboardPage() {
  const { user, loading: authLoading } = useAuth();
  const router = useRouter();
  const pathname = usePathname();
  const params = useSearchParams();

  const [bookings, setBookings] = useState<Booking[]>([]);
  const [bookingRequests, setBookingRequests] = useState<BookingRequest[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState("");

  const initialTab = params.get("tab") === "bookings" ? "bookings" : "requests";
  const [activeTab, setActiveTab] = useState<"requests" | "bookings">(initialTab);

  useEffect(() => {
    const params = new URLSearchParams(window.location.search);
    params.set("tab", activeTab);
    router.replace(`${pathname}?${params.toString()}`);
  }, [activeTab, router, pathname]);

  useEffect(() => {
    if (authLoading) return;
    if (!user) {
      router.push(`/login?next=${encodeURIComponent(pathname)}`);
      return;
    }
    if (user.user_type !== "client") {
      router.push("/dashboard/artist");
      return;
    }

    const fetchData = async () => {
      try {
        const [bookingsData, requestsData] = await Promise.all([
          getMyClientBookings(),
          getMyBookingRequests(),
        ]);
        setBookings(bookingsData.data);
        setBookingRequests(requestsData.data);
      } catch (err) {
        console.error("Client dashboard fetch error:", err);
        setError("Failed to load dashboard data. Please try again.");
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [user, authLoading, router, pathname]);

  const upcomingBookings = useMemo(() => {
    const now = new Date().getTime();
    return bookings
      .filter((b) => new Date(b.start_time).getTime() >= now)
      .sort((a, b) => new Date(a.start_time).getTime() - new Date(b.start_time).getTime())
      .slice(0, 5);
  }, [bookings]);

  if (loading) {
    return (
      <MainLayout>
        <div className="p-8 flex justify-center"><Spinner /></div>
      </MainLayout>
    );
  }

  if (error) {
    return (
      <MainLayout>
        <div className="p-8 text-red-600">{error}</div>
      </MainLayout>
    );
  }

  return (
    <MainLayout>
      <div className="max-w-5xl mx-auto px-4 py-8">
        <DashboardTabs activeTab={activeTab} onChange={setActiveTab} showServices={false} />
        <div className="mt-6">
          {activeTab === "requests" && (
            <section>
              <SectionList
                title="Booking Requests"
                data={bookingRequests}
                renderItem={(r) => <BookingRequestCard key={r.id} request={r} onUpdate={() => {}} />}
                emptyState={<span>No requests yet</span>}
              />
            </section>
          )}
          {activeTab === "bookings" && (
            <section>
              <SectionList
                title="Upcoming Bookings"
                data={upcomingBookings}
                emptyState={<span>No bookings yet</span>}
                renderItem={(booking) => (
                  <div key={booking.id} className="bg-white p-4 shadow rounded-lg">
                    <div className="font-medium text-gray-900">
                      {booking.artist?.first_name || "Unknown"} {booking.artist?.last_name || ""}
                    </div>
                    <div className="text-sm text-gray-500">{booking.service?.title || "â€”"}</div>
                    <div className="text-sm text-gray-500">
                      {format(new Date(booking.start_time), "MMM d, yyyy h:mm a")}
                    </div>
                    <div className="mt-2 flex justify-between items-center">
                      <span
                        className={`inline-flex rounded-full px-2 text-xs font-semibold leading-5 ${
                          booking.status === "completed"
                            ? "bg-green-100 text-green-800"
                            : booking.status === "cancelled"
                            ? "bg-red-100 text-red-800"
                            : booking.status === "confirmed"
                            ? "bg-brand-light text-brand-dark"
                            : "bg-yellow-100 text-yellow-800"
                        }`}
                      >
                        {formatStatus(booking.status)}
                      </span>
                      <span className="text-sm text-gray-500">
                        {formatCurrency(Number(booking.total_price))}
                      </span>
                    </div>
                  </div>
                )}
              />
              {bookings.length > upcomingBookings.length && (
                <div className="mt-2">
                  <Link href="/dashboard/client/bookings" className="text-brand-dark hover:underline text-sm">
                    View All Bookings
                  </Link>
                </div>
              )}
            </section>
          )}
        </div>
      </div>
    </MainLayout>
  );
}

