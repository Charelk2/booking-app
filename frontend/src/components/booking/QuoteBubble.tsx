import React from 'react';
import clsx from 'clsx';
import { formatCurrency } from '@/lib/utils';
import StatusBadge from '../ui/StatusBadge';

export interface QuoteBubbleProps {
  description: string;
  price: number;
  soundFee?: number;
  travelFee?: number;
  accommodation?: string;
  discount?: number;
  subtotal: number;
  total: number;
  status: 'Pending' | 'Accepted' | 'Rejected';
  className?: string;
}

export default function QuoteBubble({
  description,
  price,
  soundFee,
  travelFee,
  accommodation,
  discount,
  subtotal,
  total,
  status,
  className,
}: QuoteBubbleProps) {
  return (
    <div
      className={clsx(
        'bg-brand/10 text-brand-dark rounded-xl p-4 max-w-xs sm:max-w-md dark:bg-brand-dark/30 dark:text-brand-light',
        className,
      )}
    >
      <h4 className="text-sm font-semibold text-brand-dark dark:text-brand-light mb-1">
        Quote
      </h4>
      <ul className="space-y-1">
        <li className="text-sm text-gray-800 dark:text-gray-200">
          {description} â€“{' '}
          <span className="font-semibold">
            {formatCurrency(Number(price))}
          </span>
        </li>
        {soundFee !== undefined && (
          <li className="text-sm text-gray-800 dark:text-gray-200">
            Sound fee:{' '}
            <span className="font-semibold">
              {formatCurrency(Number(soundFee))}
            </span>
          </li>
        )}
        {travelFee !== undefined && (
          <li className="text-sm text-gray-800 dark:text-gray-200">
            Travel fee:{' '}
            <span className="font-semibold">
              {formatCurrency(Number(travelFee))}
            </span>
          </li>
        )}
        {accommodation && (
          <li className="text-sm text-gray-800 dark:text-gray-200">
            Accommodation: {accommodation}
          </li>
        )}
        {discount !== undefined && (
          <li className="text-sm text-gray-800 dark:text-gray-200">
            Discount:{' '}
            <span className="font-semibold">
              {formatCurrency(Number(discount))}
            </span>
          </li>
        )}
        <li className="text-sm text-gray-800 dark:text-gray-200">
          Subtotal:{' '}
          <span className="font-semibold">
            {formatCurrency(Number(subtotal))}
          </span>
        </li>
        <li className="text-sm text-gray-800 dark:text-gray-200">
          Total:{' '}
          <span className="font-semibold">
            {formatCurrency(Number(total))}
          </span>
        </li>
      </ul>
      <div className="mt-2">
        <StatusBadge status={status} />
      </div>
    </div>
  );
}
