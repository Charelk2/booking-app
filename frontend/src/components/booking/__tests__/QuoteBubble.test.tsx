import { createRoot } from 'react-dom/client';
import React from 'react';
import { act } from 'react-dom/test-utils';
import QuoteBubble from '../QuoteBubble';

describe('QuoteBubble', () => {
  it('matches snapshot', () => {
    const div = document.createElement('div');
    const root = createRoot(div);

    act(() => {
      root.render(
        <QuoteBubble
          description="Performance"
          price={100}
          soundFee={10}
          travelFee={20}
          accommodation="Hotel"
          discount={5}
          subtotal={125}
          total={130}
          status="Pending"
        />,
      );
    });

    expect(div.firstChild).toMatchSnapshot();

    act(() => {
      root.unmount();
    });
    div.remove();
  });
});
