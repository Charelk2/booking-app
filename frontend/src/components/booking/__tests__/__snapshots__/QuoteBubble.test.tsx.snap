// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuoteBubble matches snapshot 1`] = `
<div
  class="bg-brand/10 text-brand-dark rounded-xl p-4 max-w-xs sm:max-w-md dark:bg-brand-dark/30 dark:text-brand-light"
>
  <h4
    class="text-sm font-semibold text-brand-dark dark:text-brand-light mb-1"
  >
    Quote
  </h4>
  <ul
    class="space-y-1"
  >
    <li
      class="text-sm text-gray-800 dark:text-gray-200"
    >
      Performance
       –
       
      <span
        class="font-semibold"
      >
        R 100,00
      </span>
    </li>
    <li
      class="text-sm text-gray-800 dark:text-gray-200"
    >
      Sound fee:
       
      <span
        class="font-semibold"
      >
        R 10,00
      </span>
    </li>
    <li
      class="text-sm text-gray-800 dark:text-gray-200"
    >
      Travel fee:
       
      <span
        class="font-semibold"
      >
        R 20,00
      </span>
    </li>
    <li
      class="text-sm text-gray-800 dark:text-gray-200"
    >
      Accommodation: 
      Hotel
    </li>
    <li
      class="text-sm text-gray-800 dark:text-gray-200"
    >
      Discount:
       
      <span
        class="font-semibold"
      >
        R 5,00
      </span>
    </li>
    <li
      class="text-sm text-gray-800 dark:text-gray-200"
    >
      Subtotal:
       
      <span
        class="font-semibold"
      >
        R 125,00
      </span>
    </li>
    <li
      class="text-sm text-gray-800 dark:text-gray-200"
    >
      Total:
       
      <span
        class="font-semibold"
      >
        R 130,00
      </span>
    </li>
  </ul>
  <div
    class="mt-2"
  >
    <span
      class="inline-flex rounded-full px-2 py-0.5 text-xs font-medium bg-yellow-100 text-yellow-800"
    >
      Pending
    </span>
  </div>
</div>
`;
